src/BleOtaUploader.cpp

```diff
@@ -1,288 +1,499 @@
+        handlePackageInd(BleOtaPackageInd{data, size});
         break;
-    case PACKAGE:
-        handlePackage(data + 1, length - 1);
+    case BleOtaHeader::EndReq:
+        BleOtaEndReq::isValidSize(size) ?
+            handleEndReq(BleOtaEndReq{data}) :
+            handleError(BleOtaStatus::IncorrectMessageSize);
         break;
-    case END:
-        handleEnd(data + 1, length - 1);
+    case BleOtaHeader::SignatureReq:
+        handleSignatureReq(BleOtaSignatureReq{data, size});
         break;
-    case SET_PIN_CODE:
-        handleSetPinCode(data + 1, length - 1);
+    case BleOtaHeader::SetPinReq:
+        BleOtaSetPinReq::isValidSize(size) ?
+            handleSetPinReq(BleOtaSetPinReq{data}) :
+            handleError(BleOtaStatus::IncorrectMessageSize);
         break;
-    case REMOVE_PIN_CODE:
-        handleRemovePinCode(data + 1, length - 1);
+    case BleOtaHeader::RemovePinReq:
+        BleOtaRemovePinReq::isValidSize(size) ?
+            handleRemovePinReq(BleOtaRemovePinReq{}) :
+            handleError(BleOtaStatus::IncorrectMessageSize);
         break;
     default:
-        handleError(INCORRECT_FORMAT);
+        handleError(BleOtaStatus::IncorrectMessageHeader);
         break;
     }
 }
 
-void BleOtaUploader::handleBegin(const uint8_t* data, size_t length)
+bool BleOtaUploader::setSignatureKey(const char* key, size_t size)
 {
-    if (uploading)
-        terminateUpload();
+    return _signature.setPublicKey(key, size);
+}
+
+void BleOtaUploader::setPinCallbacks(BleOtaPinCallbacks& cb)
+{
+    _pinCallbacks = &cb;
+}
+
+void BleOtaUploader::setUploadCallbacks(BleOtaUploadCallbacks& cb)
+{
+    _uploadCallbacks = &cb;
+}
 
-    if (not enabled)
+void BleOtaUploader::handleInitReq(const BleOtaInitReq& req)
+{
+    BLE_OTA_LOG(TAG, "Handle InitReq");
+
+    const bool canCompress = _decompressor.isSupported();
+    const bool canChecksum = _checksum.isSupported();
+    const bool canUpload = _state != State::Disable;
+    const bool canSignature = _signature.isEnabled();
+    const bool canPin = _pinCallbacks != &dummyPinCallbacks;
+
+    BLE_OTA_LOG(TAG, "Send InitResp: "
+        "flags: (compr: %u, checksum: %u, upload: %u, sig: %u, pin: %u)",
+        canCompress, canChecksum, canUpload, canSignature, canPin);
+
+    sendMessage(BleOtaInitResp{{
+        .compression = canCompress,
+        .checksum = canChecksum,
+        .upload = canUpload,
+        .signature = canSignature,
+        .pin = canPin
+    }});
+}
+
+void BleOtaUploader::handleBeginReq(const BleOtaBeginReq& req)
+{
+    BLE_OTA_LOG(TAG, "Handle BeginReq: "
+        "size: (fw: %lu, pkg: %lu, buff: %lu, compr: %lu), "
+        "flags: (compr: %u, checksum: %u)",
+        req.firmwareSize, req.packageSize, req.bufferSize, req.compressedSize,
+        req.flags.compression, req.flags.checksum);
+
+    if (_state == State::Upload)
     {
-        send(UPLOAD_DISABLED);
-        return;
+        BLE_OTA_LOG(TAG, "Terminate");
```

Comment:
shall we add some info why terminated?

Reply:
Hmm, it's only one place where we have termination without error, just because upload restarted before end. Other are terminates are with error. Not sure if we need add something more. Any proposal?

Reply:
Maybe `"Restart"` ?

Reply:
Changed https://github.com/vovagorodok/ArduinoBleOTA/pull/48/commits/0130e7cb1687fbfb08e341f68af947338a12c5af